Add wear bar color API (#13328)
minetest/minetest: doc/lua_api.md

Wear Bar Color
==============

'Wear Bar' is a property of items that defines the coloring
of the bar that appears under damaged tools.

> If it is absent, the old behavior of green, yellow, red is
> used.

Reviewer: ```suggestion
If it is absent, the old behavior of green-yellow-red is
used.
```

Wear bar colors definition
--------------------------

### Syntax

For non-blending mode:
```lua

> {
>     {
>         color="#ff00ff",
>         min_durability=0.2, -- minimum durability is inclusive
>         max_durability=0.3  -- maximum durability is exclusive
>     },
>     {
>         color="#c0ffee",
>         min_durability=0.45,
>         max_durability=0.6
>     },
>     default="#ffff00",      -- color to use if no other ranges match the durability
>     blend=false
> }

Reviewer: Add spaces around operator. If the comment is too long, move it above the line.
```suggestion
{
    {
        color = "#ff00ff",
        min_durability = 0.2, -- inclusive
        max_durability = 0.3  -- exclusive
    },
    {
        color = "#c0ffee",
        min_durability = 0.45,
        max_durability = 0.6
    },
    -- color to use if no other ranges match the durability
    default = "#ffff00",
    blend = false
}
```

```

For blending mode:
```lua

> {
>     [0.2] = "#ff00ff",      -- specify color for a specific durability percent, and blend between them
>     [0.45] = "#c0ffee",
>     default="#ffff00",      -- used for 0.0% and 100.0% durability if no color explicitly specified
>     blend=true
> }

Reviewer: Same with this block.
```suggestion
{
	-- specify color for a specific durability percent, and blend between them
    [0.2] = "#ff00ff",
    [0.45] = "#c0ffee",
	-- used for 0% and 100% durability if no color explicitly specified
    default = "#ffff00",
    blend = true
}
```

```

> ### Blend Mode `blend`
> 
> When true, blends smoothly between each defined color point.
> When false, no interpolation is used, and instead colors are defined for ranges of durabilities.
> `blend` is optional and defaults false.

Reviewer: Please check the capitalization of the subchapter's title. To tell more than one cases, use bullet list.
```suggestion
### Blend mode `blend`

* If true, it blends smoothly between each defined color point.
* If false, no interpolation is used and instead colors are defined for ranges of durabilities.
`blend` is optional and defaults false.
```

> ### Default Color `default`
> 
> When `blend` is false, the color to use if the percentage of remaining durability does not match any values.
> When `blend` is true, the color to use on the 'outside' of the range of colors, if no such color is defined

Reviewer: Same with this block.
```suggestion
### Default color `default`

* If `blend` is false, the color to use if the percentage of remaining durability does not match any values.
* If `blend` is true, the color to use on the outside of the range of colors if no such color is defined.
```

### Color values

> In blending mode, specified as `float` keys assigned to a `ColorString` values.
> In non-blending mode, specified as tables containing the following keys:

Reviewer: Split to two paragraphs.
```suggestion
In blending mode, specified as `float` keys assigned to a `ColorString` values.

In non-blending mode, specified as tables containing the following keys:
```

* `min_durability`: `float`, minimum durability to show color at (inclusive)
* `max_durability`: `float`, maximum durability to show color at (exclusive)

> * `color`:          `ColorString`, color to use for bar in the specified range

Reviewer: There is no need to align for this value (with lines above) as it will be shown the same as a Markdown file.
```suggestion
* `color`: `ColorString`, color to use for bar in the specified range
```

Wear Bar Color
==============

'Wear Bar' is a property of items that defines the coloring
of the bar that appears under damaged tools.
If it is absent, the old behavior of green-yellow-red is
used.

Wear bar colors definition
--------------------------

### Syntax

For non-blending mode:
```lua
{
    {
        color = "#ff00ff",
        min_durability = 0.2, -- inclusive
        max_durability = 0.3  -- exclusive
    },
    {
        color = "#c0ffee",
        min_durability = 0.45,
        max_durability = 0.6
    },
    -- color to use if no other ranges match the durability
    default = "#ffff00",
    blend = false
}
```

For blending mode:
```lua
{

> 	-- specify color for a specific durability percent, and blend between them
>     [0.2] = "#ff00ff",
>     [0.45] = "#c0ffee",
> 	-- used for 0% and 100% durability if no color explicitly specified
>     default = "#ffff00",
>     blend = true

Reviewer: Sorry, I made a mistake when writing the suggestion. It should be all spaces (4 spaces for one indentation).
```suggestion
    -- specify color for a specific durability percent, and blend between them
    [0.2] = "#ff00ff",
    [0.45] = "#c0ffee",
    -- used for 0% and 100% durability if no color explicitly specified
    default = "#ffff00",
    blend = true
```

Wear Bar Color
==============

'Wear Bar' is a property of items that defines the coloring
of the bar that appears under damaged tools.
If it is absent, the old behavior of green-yellow-red is

> used.

Reviewer 2: calling it "default behavior" would make more sense, it's not like we're going to remove it

Wear bar colors definition
--------------------------

### Syntax

```lua
{
    -- 'constant' or 'linear'
    -- (nil defaults to 'constant')
    blend = "linear",
    color_stops = {
        [0.0] = "#ff0000",
        [0.5] = "slateblue",
        [1.0] = {r=0, g=255, b=0, a=150},
    }
}
```

### Blend mode `blend`

* `linear`: blends smoothly between each defined color point.
* `constant`: each color starts at its defined point, and continues up to the next point

### Color stops `color_stops`

> Specified as `float` keys assigned to `ColorSpec` values.

Reviewer 2: ```suggestion
Specified as `ColorSpec` values assigned to `float` keys.
```

One thing that's odd about the API design: The keys in `color_stops` are not wear values, but durability values (`1 - wear`). This is not documented and inconsistent with the rest of the API. For consistency, it would make more sense to use wear values.

        -- fallback behavior.
    },

>     -- See "Wear Bar Color" section for an example including explanation
>         wear_color = {
>             {
>                 color="#ff00ff",
>                 min_durability=0.2,
>                 max_durability=0.3
>             },
>             {
>                 color="#c0ffee",
>                 min_durability=0.45,
>                 max_durability=0.6
>             },
>             default="#ffff00",
>             blend=false
>         }

Reviewer: Check the indentation. Add simple description before referencing to the section.
```suggestion
    -- Set wear bar color of the tool by setting color points or ranges
    -- See "Wear Bar Color" section for further explanation including example
    wear_color = {
        {
            color = "#ff00ff",
            min_durability = 0.2,
            max_durability = 0.3
        },
        {
            color = "#c0ffee",
            min_durability = 0.45,
            max_durability = 0.6
        },
        default = "#ffff00",
        -- default color for fallback
        blend = false
        -- whether to blend the colors
    }
```

        node_placement_prediction = nil,
        -- If nil and item is node, prediction is made automatically.
        -- If nil and item is not a node, no prediction is made.
        -- If "" and item is anything, no prediction is made.
        -- Otherwise should be name of node which the client immediately places
        -- on ground when the player places the item. Server will always update

>         -- with actual result shortly.

Reviewer: Please check indentation with this block, too.

        -- fallback behavior.
    },

    -- Set wear bar color of the tool by setting color points or ranges
    -- See "Wear Bar Color" section for further explanation including example

>     wear_color = {

Author: done.
